File: Dynamic Programming/Hard/LET_no132_Palindrome Partitioning II.cpp
class Solution {
public:
    int minCut(string s) {
        int n = s.size();
        vector<int> res;
        for (int i = 0; i <= n; i++)
            res.push_back(i - 1);
        
        vector<vector<bool>> isPalindrome = getPalindrome(s);
        
        for (int length = 1; length <= n; length++) {
            for (int i = 0; i < length; i++) {
                if (isPalindrome[i][length - 1])
                    res[length] = min(res[length], res[i] + 1);
            }
        }
        
        return res[n];
    }
    
    vector<vector<bool>> getPalindrome(string s) {
        int n = s.size();
        vector<vector<bool>> res(n, vector<bool>(n, false));
            
        for (int i = 0; i < n; i++) {
            res[i][i] = true;
            if (i < n - 1)
                res[i][i + 1] = s[i] == s[i + 1];
        }
        
        for (int d = 2; d < n; d++) {
            for (int i = 0; i < n - d; i++)
                res[i][i + d] = res[i + 1][i + d - 1] && s[i] == s[i + d];
        }
        
        return res;
    }
};

File: Dynamic Programming/Hard/LET_no132_Palindrome Partitioning II.py
class Solution:
    """
    @param s: A string
    @return: An integer
    """

    def minCut(self, s):
        if not s:
            return 0
        
        n = len(s)
        is_palindrome = self.getPalindrome(s, n)
        cuts = [i - 1 for i in range(n + 1)]
        
        for length in range(1, n + 1):
            for left in range(length):
                right = length - 1
                if is_palindrome[left][right]:
                    cuts[length] = min(cuts[length], cuts[left] + 1)
        
        return cuts[n]
    
    def getPalindrome(self, s, n):
        dp = [[False] * n for _ in range(n)]
        
        for length in range(n):
            for left in range(n - length):
                right = left + length
                dp[left][right] = s[left] == s[right] and (right - left < 2 or dp[left + 1][right - 1])
                
        return dp
